### PR TITLE
Detect NLS pattern

### PR DESCRIPTION
**Pull request by tmigot**
## Related issues

Closes #

## Checklist

- [x] I am following the [contributing guidelines](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/docs/src/90-contributing.md)
- [x] Tests are passing
- [x] Lint workflow is passing
- [x] Docs were updated and workflow is passing

### CODE REVIEW

File: src/parse-functions.jl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Check if the given model has a nonlinear least squares objective.
+
+The function uses the package ExpressionTreeForge to get the expression tree of the objective function,
+then try to detect the least squares pattern. There is no guarantee that this function detects it accurately.
+"""
+function isnls(nlp)
+  expr_tree = ExpressionTreeForge.get_expression_tree(nlp)
+  F_expr = extract_element_functions(expr_tree)
+  test_square(expr) = expr.field == ExpressionTreeForge.M_power_operator.Power_operator{Int}(2)
```
Comment (paraynaud) [inline src/parse-functions.jl:18]: Is `Int` in `Power_operator{Int}(2)` limiting in practice ?

File: src/parse-functions.jl [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  myand(a::Bool,b::Bool) = a && b
+  is_nls = mapreduce(test_square, myand, F_expr) :: Bool
```
Comment (paraynaud) [inline src/parse-functions.jl:20]: I was using a lot the `myand(a,b)` in the past, but I happened to discover that `&` replaces very well `myand`.
Something like
```suggestion
  is_nls = mapreduce(test_square, &, F_expr) :: Bool
```
should work.

File: src/parse-functions.jl [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#using ExpressionTreeForge
+#using ..ExpressionTreeForge.M_power_operator
+
+"""
+    isnlp(nlp::ADNLPModels.AbstractADNLPModel)
+    isnlp(nlp::MathOptInterface.Nonlinear.Model)
+    isnlp(nlp::JuMP.Model)
+    isnlp(nlp::MathOptNLPModel)
+
+Check if the given model has a nonlinear least squares objective.
```
Comment (dpo) [inline src/parse-functions.jl:10]: Why is it called `isnlp` then? 😄

File: src/parse-functions.jl (hunk 1 above)
Comment (dpo) [reply src/parse-functions.jl:18]: I wouldn’t think so.

**Comment by dpo** [comment]
This is really awesome @tmigot ! 👍

**Comment by tmigot** [comment]
Thanks @paraynaud !

**Comment by paraynaud** [comment]
My pleasure :)

**Pull request merged**